File: content/snippets/go/language_basics_interfaces_and_polymorphism.go
package main

import "fmt"

type Speaker interface {
	Speak() string
}

type Dog struct{}

type Cat struct{}

func (Dog) Speak() string {
	return "woof"
}

func (Cat) Speak() string {
	return "meow"
}

func main() {
	speakers := []Speaker{Dog{}, Cat{}}
	for _, speaker := range speakers {
		fmt.Println(speaker.Speak())
	}
}

File: content/snippets/python/language_basics_interfaces_and_polymorphism.py
#!/usr/bin/env python3
from abc import ABC, abstractmethod


class Speaker(ABC):
    @abstractmethod
    def speak(self) -> str:
        raise NotImplementedError


class Dog(Speaker):
    def speak(self) -> str:
        return "woof"


class Cat(Speaker):
    def speak(self) -> str:
        return "meow"


def main() -> None:
    speakers: list[Speaker] = [Dog(), Cat()]
    for speaker in speakers:
        print(speaker.speak())


if __name__ == "__main__":
    main()
